File: backend/agents/validation_agent.py
"""Recipe validation agent."""


class ValidationAgent:
    ALLOWED_UNITS = {"g", "ml", "unit"}

    def validate_recipe(self, recipe: dict) -> dict:
        if not recipe.get("title"):
            raise ValueError("Recipe title is required.")
        if not recipe.get("ingredients"):
            raise ValueError("Recipe ingredients are required.")
        if not recipe.get("instructions"):
            raise ValueError("Recipe instructions are required.")

        servings = recipe.get("servings", 1)
        if not isinstance(servings, int) or servings <= 0:
            raise ValueError("Recipe servings must be a positive integer.")

        for ingredient in recipe["ingredients"]:
            unit = str(ingredient.get("unit", "")).strip().lower()
            if unit not in self.ALLOWED_UNITS:
                raise ValueError(f"Invalid ingredient unit: {unit}")

        return recipe

File: supabase/functions/api/agents/validationAgent.ts
export function validateRecipe(recipe: Record<string, unknown>): Record<string, unknown> {
  if (!recipe.title || typeof recipe.title !== "string") throw new Error("Recipe title is required.");
  if (!Array.isArray(recipe.ingredients) || recipe.ingredients.length === 0) throw new Error("Recipe ingredients are required.");
  if (!Array.isArray(recipe.instructions) || recipe.instructions.length === 0) throw new Error("Recipe instructions are required.");

  const servings = Number(recipe.servings);
  if (!Number.isInteger(servings) || servings <= 0) throw new Error("Recipe servings must be a positive integer.");

  const allowed = new Set(["g", "ml", "unit"]);
  for (const ing of recipe.ingredients as Array<Record<string, unknown>>) {
    if (!ing.name || !ing.unit || ing.amount === undefined) throw new Error("Each ingredient must include name, amount, and unit.");
    if (!allowed.has(String(ing.unit))) throw new Error("Ingredient units must be g, ml, or unit.");
  }
  return recipe;
}
